File: backend/app/models/trading_plan.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, Numeric
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base


class TradingPlan(Base):
    """매매계획 이력 (계획 저장 + 체결 기록)"""
    __tablename__ = "trading_plans"

    id = Column(Integer, primary_key=True, index=True)
    position_id = Column(Integer, ForeignKey("positions.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    version = Column(Integer, default=1)

    # 이력 유형: plan_saved (계획 저장) | execution (체결 기록)
    record_type = Column(String(20), default='plan_saved')

    # === 계획 저장 (record_type='plan_saved') ===
    buy_plan = Column(JSON, nullable=True)
    take_profit_targets = Column(JSON, nullable=True)
    stop_loss_targets = Column(JSON, nullable=True)
    memo = Column(Text, nullable=True)

    # 변경 이력 (AI 컨텍스트용 JSON) - deprecated, 이제 개별 레코드로 관리
    changes = Column(JSON, nullable=True)

    # === 체결 기록 (record_type='execution') ===
    # 체결 유형: buy | take_profit | stop_loss
    plan_type = Column(String(20), nullable=True)
    # 몇 번째 체결인지 (1차 익절, 2차 익절 등)
    execution_index = Column(Integer, nullable=True)
    # 계획했던 가격/수량
    target_price = Column(Numeric(20, 8), nullable=True)
    target_quantity = Column(Numeric(20, 8), nullable=True)
    # 실제 체결 가격/수량
    executed_price = Column(Numeric(20, 8), nullable=True)
    executed_quantity = Column(Numeric(20, 8), nullable=True)
    # 체결 금액
    executed_amount = Column(Numeric(20, 2), nullable=True)
    # 실현 손익 (익절/손절만)
    profit_loss = Column(Numeric(20, 2), nullable=True)
    profit_rate = Column(Numeric(10, 6), nullable=True)

    # 상태
    status = Column(String(20), default='draft')  # draft, submitted

    created_at = Column(DateTime, default=datetime.utcnow)
    submitted_at = Column(DateTime, nullable=True)

    # 관계
    position = relationship("Position", back_populates="trading_plans")
    user = relationship("User", back_populates="trading_plans")

File: spring-backend/backend/src/main/java/com/fundmessenger/trading/entity/TradingPlan.java
package com.fundmessenger.trading.entity;

import com.fundmessenger.position.entity.Position;
import com.fundmessenger.user.entity.User;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.annotations.Type;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Entity
@Table(name = "trading_plans")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class TradingPlan {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "position_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    private Position position;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "version", columnDefinition = "integer default 1")
    private Integer version = 1;

    @Column(name = "record_type", length = 20, columnDefinition = "varchar(20) default 'plan_saved'")
    private String recordType = "plan_saved";

    @Type(JsonType.class)
    @Column(name = "buy_plan", columnDefinition = "jsonb")
    private Object buyPlan;

    @Type(JsonType.class)
    @Column(name = "take_profit_targets", columnDefinition = "jsonb")
    private Object takeProfitTargets;

    @Type(JsonType.class)
    @Column(name = "stop_loss_targets", columnDefinition = "jsonb")
    private Object stopLossTargets;

    @Column(name = "memo", columnDefinition = "text")
    private String memo;

    @Type(JsonType.class)
    @Column(name = "changes", columnDefinition = "jsonb")
    private Object changes;

    @Column(name = "plan_type", length = 20)
    private String planType;

    @Column(name = "execution_index")
    private Integer executionIndex;

    @Column(name = "target_price", precision = 20, scale = 8)
    private BigDecimal targetPrice;

    @Column(name = "target_quantity", precision = 20, scale = 8)
    private BigDecimal targetQuantity;

    @Column(name = "executed_price", precision = 20, scale = 8)
    private BigDecimal executedPrice;

    @Column(name = "executed_quantity", precision = 20, scale = 8)
    private BigDecimal executedQuantity;

    @Column(name = "executed_amount", precision = 20, scale = 2)
    private BigDecimal executedAmount;

    @Column(name = "profit_loss", precision = 20, scale = 2)
    private BigDecimal profitLoss;

    @Column(name = "profit_rate", precision = 10, scale = 6)
    private BigDecimal profitRate;

    @Column(name = "status", length = 20, columnDefinition = "varchar(20) default 'draft'")
    private String status = "draft";

    @CreationTimestamp
    @Column(name = "created_at", columnDefinition = "timestamptz")
    private OffsetDateTime createdAt;

    @Column(name = "submitted_at", columnDefinition = "timestamptz")
    private OffsetDateTime submittedAt;
}
